sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/SpecializedGettersReader.java

```diff
@@ -0,0 +1,73 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.expressions;
+
+import org.apache.spark.sql.types.*;
+
+public class SpecializedGettersReader {
+  private final boolean handleNull;
+  private final boolean handleUserDefinedType;
+
+  public SpecializedGettersReader(boolean handleNull, boolean handleUserDefinedType) {
+    this.handleNull = handleNull;
+    this.handleUserDefinedType = handleUserDefinedType;
+  }
+
+  public Object read(SpecializedGetters obj, int ordinal, DataType dataType) {
+    if (handleNull && (obj.isNullAt(ordinal) || dataType instanceof NullType)) {
+      return null;
+    } else if (dataType instanceof BooleanType) {
+      return obj.getBoolean(ordinal);
+    } else if (dataType instanceof ByteType) {
+      return obj.getByte(ordinal);
+    } else if (dataType instanceof ShortType) {
+      return obj.getShort(ordinal);
+    } else if (dataType instanceof IntegerType) {
+      return obj.getInt(ordinal);
+    } else if (dataType instanceof LongType) {
+      return obj.getLong(ordinal);
+    } else if (dataType instanceof FloatType) {
+      return obj.getFloat(ordinal);
+    } else if (dataType instanceof DoubleType) {
+      return obj.getDouble(ordinal);
+    } else if (dataType instanceof DecimalType) {
+      DecimalType dt = (DecimalType) dataType;
+      return obj.getDecimal(ordinal, dt.precision(), dt.scale());
+    } else if (dataType instanceof DateType) {
+      return obj.getInt(ordinal);
+    } else if (dataType instanceof TimestampType) {
+      return obj.getLong(ordinal);
+    } else if (dataType instanceof BinaryType) {
```

Comment:
The order of evaluating Binary/String/DecimalType varied in the code this replaces. That shouldn't matter right -- none of these types are subtypes of the other.

I wonder, while we're here, is it sensible to put checks for more common types further up? for example, 'short' is pretty rare. String is probably more common than binary or decimal. Those are all maybe more common that date types. Etc. I understand for readability we might not want to move short, but might be worth rearranging the others for some performance.

Reply:
IMHO, for performance, yes we can move commonly used types to top of huge if statements (though I don't have any statistic for this - we all know string should be placed to one of top), and for readability, this looks to be thoughtfully placed - primitive types first, dates, binary, string, complex types (`CalendarInterval` could be moved to dates group).

Reply:
No, I have nothing but intuition here. I wouldn't make major changes. I think we can probably make the obvious ones, like, String should probably follow primitives.

Reply:
OK. That's also intuitive to me. Will address. Please let me know if there're other obvious things.